drivers/display/ls0xx.c

```diff
@@ -0,0 +1,352 @@
+ * +-------------------+-------------------+----------------+
+ */
+#define LS0XX_BYTES_PER_LINE  ((LS0XX_PANEL_WIDTH / LS0XX_PIXELS_PER_BYTE) + 2)
+
+#define LS0XX_BIT_WRITECMD    0x01
+#define LS0XX_BIT_VCOM        0x02
+#define LS0XX_BIT_CLEAR       0x04
+
+static uint8_t line_buf[LS0XX_BYTES_PER_LINE];
+
+static struct spi_buf tx_buf;
+static struct spi_buf_set buf_set = { .buffers = &tx_buf, .count = 1 };
+
+struct ls0xx_data {
+#if DT_INST_NODE_HAS_PROP(0, disp_en_gpios)
+	const struct device *disp_dev;
+#endif
+#if DT_INST_NODE_HAS_PROP(0, extcomin_gpios)
+	const struct device *extcomin_dev;
+#endif
+	const struct device *spi_dev;
+	struct spi_cs_control cs_ctrl;
+	struct spi_config spi_config;
+};
+
+
+#if DT_INST_NODE_HAS_PROP(0, extcomin_gpios)
+/* Driver will handle VCOM toggling */
+static void ls0xx_vcom_toggle(void *a, void *b, void *c)
+{
+	struct ls0xx_data *driver = (struct ls0xx_data *)a;
+
+	while (1) {
+		gpio_pin_toggle(driver->extcomin_dev,
+				DT_INST_GPIO_PIN(0, extcomin_gpios));
+		k_usleep(3);
+		gpio_pin_toggle(driver->extcomin_dev,
+				DT_INST_GPIO_PIN(0, extcomin_gpios));
+		k_msleep(1000 / DT_INST_PROP(0, extcomin_frequency));
+	}
+}
+
+K_THREAD_STACK_DEFINE(vcom_toggle_stack, 256);
+struct k_thread vcom_toggle_thread;
+#endif
+
+static int ls0xx_blanking_off(const struct device *dev)
+{
+#if DT_INST_NODE_HAS_PROP(0, disp_en_gpios)
+	struct ls0xx_data *driver = dev->data;
+
+	return gpio_pin_set(driver->disp_dev,
+			    DT_INST_GPIO_PIN(0, disp_en_gpios), 1);
+#else
+	LOG_WRN("Unsupported");
+	return -ENOTSUP;
+#endif
+}
+
+static int ls0xx_blanking_on(const struct device *dev)
+{
+#if DT_INST_NODE_HAS_PROP(0, disp_en_gpios)
+	struct ls0xx_data *driver = dev->data;
+
+	return gpio_pin_set(driver->disp_dev,
+			    DT_INST_GPIO_PIN(0, disp_en_gpios), 0);
+#else
+	LOG_WRN("Unsupported");
+	return -ENOTSUP;
+#endif
+}
+
+static int ls0xx_clear(const struct device *dev)
+{
+	struct ls0xx_data *driver = dev->data;
+	int err;
+	uint8_t clear_cmd[2] = { LS0XX_BIT_CLEAR, 0 };
+
+	tx_buf.len = sizeof(clear_cmd);
+	tx_buf.buf = &clear_cmd;
+	err = spi_write(driver->spi_dev, &driver->spi_config, &buf_set);
+	spi_release(driver->spi_dev, &driver->spi_config);
+
+	return err;
+}
+
+static int ls0xx_update_display(const struct device *dev,
+				uint16_t start_line,
+				uint16_t num_lines,
+				const uint8_t *buf)
+{
+	struct ls0xx_data *driver = dev->data;
+	int err;
+	uint8_t write_cmd;
+
+	LOG_DBG("Lines %d to %d", start_line, start_line + num_lines - 1);
+
+	write_cmd = LS0XX_BIT_WRITECMD;
+	tx_buf.len = sizeof(write_cmd);
+	tx_buf.buf =  &write_cmd;
+	err = spi_write(driver->spi_dev, &driver->spi_config, &buf_set);
+
+	if (buf != NULL) {
+		/* Send each line to the screen including
+		 * the line number and dummy bits
+		 */
+		tx_buf.len = LS0XX_BYTES_PER_LINE;
+		tx_buf.buf = &line_buf[0];
+		for (int l = start_line; l <= start_line + num_lines - 1; l++) {
+			line_buf[0] = l;
+			memcpy(&line_buf[1], buf,
```

Comment:
This memcpy and `line_buf` could be removed if two `spi_buf`s are used in `buf_set`

Reply:
That is good point. I have worked it over [quickly](https://github.com/jfischer-no/zephyr/commit/75b2d31c1c95e4b93548897387a9222b4aa67544). You are free to take it over https://github.com/jfischer-no/zephyr/commit/75b2d31c1c95e4b93548897387a9222b4aa67544.patch and squash.

Reply:
I had tried this earlier, but it affects the speed of the write. See below.
Separate buffers (suggested by you). It takes 18ms. The first transfer is the write cmd and then 3 buffers are transferred separately.
![image](https://user-images.githubusercontent.com/10092220/104937996-7a9d4c80-59d4-11eb-986f-c8f9de0bd9ed.png)

Single buffer. It takes 14ms.The first transfer is the write cmd and then the complete buffer is transferred.
![image](https://user-images.githubusercontent.com/10092220/104938149-ac161800-59d4-11eb-8026-6a9f821c54eb.png)

Both of these are tested on NRF52840 with DMA (SPIM peripheral) on LS013B7DH03. While 4ms may not seem high for a small display, it will quickly add up for a larger display and the lag would be perceivable.

Reply:
I see difference of about 0.15ms for the update of 8 lines each 144px. The benefit is that there is no need for an intermediate buffer and the code is much simpler.

![Uphie7xe](https://user-images.githubusercontent.com/9864546/105103329-bc032a00-5ab0-11eb-8a19-931d0ade61ad.png)
![feeH2qua](https://user-images.githubusercontent.com/9864546/105103338-bf96b100-5ab0-11eb-88b7-83917bfb8cba.png)

Reply:
The large difference is during updating the complete display. You can try with the example i used for toggling the complete display black and white. [(src)](https://github.com/rgujju/zephyr/blob/simple_display/samples/display/simple_display/src/main.c)

Reply:
Well, regarding complete display update, depending on how much buffer (LVGL) is available it will scale according to the display size or be a little less. Decisive for this controller is the time it takes to write one line. The delay between LINE_NUMBER and DATA/DATA and DUMMY_BYTE will have less influence the larger the number of pixels in the line is.

Whatever you decide, with or without intermediate buffers, please rework the use of `spi_buf `and `spi_buf_set` and remove global `struct spi_buf tx_buf`and `struct spi_buf_set buf_set`.